spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/repository/query/GqlDatastoreQuery.java

```diff
@@ -0,0 +1,202 @@
+			GQL_PARAM_BINDING_FUNC_MAP;
+
+	static {
+		GQL_PARAM_BINDING_FUNC_MAP = ImmutableMap
+				.<Class<?>, Function<Builder, BiFunction<String, Object, Builder>>>builder()
+				.put(Cursor.class, builder -> (s, o) -> builder.setBinding(s, (Cursor) o))
+				.put(String.class, builder -> (s, o) -> builder.setBinding(s, (String) o))
+				.put(String[].class, builder -> (s, o) -> builder.setBinding(s, (String[]) o))
+				.put(Long.class, builder -> (s, o) -> builder.setBinding(s, (Long) o))
+				.put(long[].class, builder -> (s, o) -> builder.setBinding(s, (long[]) o))
+				.put(Double.class, builder -> (s, o) -> builder.setBinding(s, (Double) o))
+				.put(double[].class, builder -> (s, o) -> builder.setBinding(s, (double[]) o))
+				.put(Boolean.class,
+						builder -> (s, o) -> builder.setBinding(s, (Boolean) o))
+				.put(boolean[].class,
+						builder -> (s, o) -> builder.setBinding(s, (boolean[]) o))
+				.put(Timestamp.class,
+						builder -> (s, o) -> builder.setBinding(s, (Timestamp) o))
+				.put(Timestamp[].class,
+						builder -> (s, o) -> builder.setBinding(s, (Timestamp[]) o))
+				.put(Key.class, builder -> (s, o) -> builder.setBinding(s, (Key) o))
+				.put(Key[].class, builder -> (s, o) -> builder.setBinding(s, (Key[]) o))
+				.put(Blob.class, builder -> (s, o) -> builder.setBinding(s, (Blob) o))
+				.put(Blob[].class, builder -> (s, o) -> builder.setBinding(s, (Blob[]) o))
+				.build();
+	}
+
+	private final QueryMethod queryMethod;
+
+	private final DatastoreOperations datastoreOperations;
+
+	protected final DatastoreMappingContext datastoreMappingContext;
+
+	private final Class<T> entityType;
+
+	private final String gql;
+
+	// unused currently, but will be used for SpEL expression in the query.
+	private QueryMethodEvaluationContextProvider evaluationContextProvider;
+
+	// unused currently, but will be used for SpEL expression in the query.
+	private SpelExpressionParser expressionParser;
+
+	/**
+	 * Constructor
+	 * @param type the underlying entity type
+	 * @param queryMethod the underlying query method to support.
+	 * @param datastoreOperations used for executing queries.
+	 * @param datastoreMappingContext used for getting metadata about entities.
+	 */
+	GqlDatastoreQuery(Class<T> type, QueryMethod queryMethod,
+			DatastoreOperations datastoreOperations, String gql,
+			QueryMethodEvaluationContextProvider evaluationContextProvider,
+			SpelExpressionParser expressionParser,
+			DatastoreMappingContext datastoreMappingContext) {
+		this.queryMethod = queryMethod;
+		this.entityType = type;
+		this.datastoreOperations = datastoreOperations;
+		this.datastoreMappingContext = datastoreMappingContext;
+		this.evaluationContextProvider = evaluationContextProvider;
+		this.expressionParser = expressionParser;
+		this.gql = StringUtils.trimTrailingCharacter(gql.trim(), ';');
+	}
+
+	@Override
+	public Object execute(Object[] parameters) {
+		List<T> rawResult = executeRawResult(parameters);
```

Comment:
Are we going to support count queries? Cassandra implementation [supports](https://github.com/spring-projects/spring-data-cassandra/blob/c843293fa52cca77d6d0860845420167664903ee/spring-data-cassandra/src/main/java/org/springframework/data/cassandra/repository/query/AbstractCassandraQuery.java#L145) it for example. Also, Datastore has count query support.

Reply:
Yup, we support it with Spanner, too. You're looking at their "Abstract" query class that supports both annotated and also name-convention query methods (that's also how I do it in Spanner). The "count" ones only come from name-convention query methods, which we aren't supporting in this PR in this file.

Reply:
Why cant you have method with `@Query` that returns a count instead of the entities?

Reply:
Unless you want to get into GQL parsing, I don't know how you could detect if a Query Method should return the domain type T or just a Long. 

The name-convention query methods' names are parsed by Spring Data into a `PartTree` for you and that will take care of checking if the name implies a `count` query method for you.

Reply:
[Cassandra](https://github.com/spring-projects/spring-data-cassandra/blob/c843293fa52cca77d6d0860845420167664903ee/spring-data-cassandra/src/main/java/org/springframework/data/cassandra/repository/Query.java) just has property on the `@Query`.
We can leave this to another PR of course.

Reply:
Can you add this as a question for our next conversation with the Spring Data team?

Reply:
Will make note.